File: src/main/java/io/github/freya022/botcommands/api/pagination/paginator/PaginatorBuilder.java
package io.github.freya022.botcommands.api.pagination.paginator;

import io.github.freya022.botcommands.api.components.Components;
import org.jetbrains.annotations.NotNull;

/**
 * Builds a {@link Paginator}
 */
public final class PaginatorBuilder extends BasicPaginatorBuilder<PaginatorBuilder, Paginator> {
	private int maxPages;

	public PaginatorBuilder(@NotNull Components componentsService) {
		super(componentsService);
	}

	public PaginatorBuilder setMaxPages(int maxPages) {
		this.maxPages = maxPages;

		return this;
	}

	@Override
	@NotNull
	public Paginator build() {
		return new Paginator(componentsService, constraints, timeout, maxPages, paginatorSupplier, hasDeleteButton, firstContent, previousContent, nextContent, lastContent, deleteContent);
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/api/pagination/paginator/PaginatorBuilder.kt
package io.github.freya022.botcommands.api.pagination.paginator

import io.github.freya022.botcommands.api.core.BContext
import io.github.freya022.botcommands.api.pagination.PageEditor
import io.github.freya022.botcommands.api.pagination.Paginators

/**
 * Builds a [Paginator].
 *
 * @see Paginators.paginator
 */
class PaginatorBuilder internal constructor(
    context: BContext,
    val pageEditor: PageEditor<Paginator>
) : BasicPaginatorBuilder<PaginatorBuilder, Paginator>(context) {
    var maxPages = 0
        private set(value) {
            check(maxPages > 0) { "Max pages must be > 0" }
            field = value
        }

    fun setMaxPages(maxPages: Int): PaginatorBuilder = config {
        this.maxPages = maxPages
    }

    override fun build(): Paginator = Paginator(context, this)
}
